addons/project_timesheet_holidays/models/hr_holidays.py

```diff
@@ -144,3 +144,21 @@ def _action_user_cancel(self, reason):
         timesheets.unlink()
         self._check_missing_global_leave_timesheets()
         return res
+
+    def _force_cancel(self, *args, **kwargs):
+        super()._force_cancel(*args, **kwargs)
+        # override this method to reevaluate timesheets after the leaves are updated via force cancel
+        timesheets = self.sudo().timesheet_ids
+        timesheets.holiday_id = False
+        timesheets.unlink()
+
+    def write(self, vals):
+        res = super().write(vals)
+        # reevaluate timesheets after the leaves are wrote in order to remove empty timesheets
+        timesheet_ids_to_remove = []
+        for leave in self:
+            if leave.number_of_days == 0 and leave.sudo().timesheet_ids:
+                leave.sudo().timesheet_ids.holiday_id = False
+                timesheet_ids_to_remove.extend(leave.timesheet_ids)
```

Comment:
Why not make a recordset instead?
```suggestion
        timesheet_ids_to_remove = self.env['account.analytic.line']
        for leave in self:
            if leave.number_of_days == 0 and leave.sudo().timesheet_ids:
                leave.sudo().timesheet_ids.holiday_id = False
                timesheet_ids_to_remove |= leave.timesheet_ids
```